buildpack/ext_descriptor.go

```diff
@@ -29,7 +31,34 @@ func ReadExtDescriptor(path string) (*ExtDescriptor, error) {
 	if descriptor.WithRootDir, err = filepath.Abs(filepath.Dir(path)); err != nil {
 		return &ExtDescriptor{}, err
 	}
-	return descriptor, nil
+	err = descriptor.inferTargets()
+	return descriptor, err
+}
+
+func (d *ExtDescriptor) inferTargets() error {
+	if len(d.Targets) == 0 {
+		binDir := filepath.Join(d.WithRootDir, "bin")
+		if stat, _ := os.Stat(binDir); stat != nil {
```

Comment:
`*` seems reasonable to me, but i wonder if it's important how consistent we keep this with the `bp_descriptor` implementation?  (right now it's a copy & paste with a few variable names changed to protect the innocent.)  I think we've been a little inconsistent / lacking-intentionality in terms of deciding what it means for the Targets to be missing vs. to be `*` - both amount to the same thing currently, but there could be different implications later...

Reply:
Ah, I missed this little `return true` and I thought missing targets don't satisfy... https://github.com/buildpacks/lifecycle/blob/28084b74124ad4484bd9e67ea4c469b05b5f38fa/platform/target_data.go#L104

I'm fine to do whatever is easiest to comprehend. But I think staying consistent across bps/exts will save us headache in the long run.

Reply:
Okay, but reading this part:
```
		if d.PlatformAPI.AtLeast("0.12") {
			targetMatch := false
			if isWildcard(d.AnalyzeMD.RunImageTarget()) || hasWildcard(descriptor.TargetsList()) {
				targetMatch = true
			} else {
				for i := range descriptor.TargetsList() {
					d.Logger.Debugf("Checking for match against descriptor:", descriptor.TargetsList()[i])
					if platform.TargetSatisfiedForBuild(*d.AnalyzeMD.RunImage.TargetMetadata, descriptor.TargetsList()[i]) {
						targetMatch = true
						break
					}
				}
			}
```
Since `descriptor.TargetsList()` is empty, we never enter the `for` loop where `targetMatch = true` and so extensions that are missing targets will not match any target. I think we need to explicitly return the wildcard...

Reply:
I'm trying to imagine the unit test that fails in this scenario, maybe you can help me sketch it?  but in general
- you have to either specify a Target or a Stack, right? 
- if they specified a stack, we should be doing matching based on the stack (i feel like this must work, given that all the old acceptance tests and building the samples repo works still)
- only if one side specified a target and no stack, then we're saying "well we can't just infer targets from nothing" and that seems... correct?  basically it means that the point at which you're not specifying stacks anymore, you're giving up on backwards compatibility, which again seems correct to me?

Reply:
> you have to either specify a Target or a Stack, right?

This makes sense for buildpacks but I think that requirement for image extensions is less clear (because executables are optional).

This might be best illustrated with an acceptance test for the detector that has an image extension with no `./bin/*` executables and an analyzed.toml that declares targets. I would expect that image extension to pass detection, but I think in the current implementation it would fail due to the empty targets list.

I think it's reasonable to infer "matches any" for image extensions that fail to declare targets because you could have a Dockerfile that only declares an ENV var or something equally permissible for all os/arch combinations.